Jetpack Installation: update selected site after the installation

hichamboushaba opened the pull request:
Closes: #5663

### Description
@hafizrahman this is just an attempt to see how we can refresh the app's status after the Jetpack installation without having to rely on Activity scoped ViewModels.

Feel free to keep it or to make any changes if you want 🙂 


On another subject, when I tested the installation, I noticed that the UI at the step of installation is not accurate, we show a checkbox before the installation si actually finished, and we show three progress indicators at the same time:

<img src="https://user-images.githubusercontent.com/1657201/150322406-7fed27d1-a793-4e0d-afb0-8aa3c3651d25.png" width=360/>

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->


- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Rever the usage of activity scoped ViewModel

hichamboushaba committed:
Check for jetpack connection after the installation

hichamboushaba committed:
Monitor Jetpack installation in MyStore and Settings screens

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/jetpack/JetpackInstallViewModel.kt:
```diff
@@ -108,7 +108,7 @@ class JetpackInstallViewModel @Inject constructor(
             if (sites != null) {
                 val syncedSite = sites.firstOrNull { it.siteId == selectedSite.get().siteId }
-                if (syncedSite?.hasWooCommerce == true) {
+                if (syncedSite?.isJetpackConnected == true && syncedSite.hasWooCommerce) {
```
We were considering the connection before it actually happens, since the app runs manual checks to update `hasWooCommerce`, its value without checking `isJetpackConnected` is not enough to consider the syncing as done.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreFragment.kt:
```diff
@@ -184,9 +184,11 @@ class MyStoreFragment :
                 is VisitorStatsViewState.Content -> showVisitorStats(stats.stats)
-                VisitorStatsViewState.Error -> binding.myStoreStats.showVisitorStatsError()
+                VisitorStatsViewState.Error -> {
+                    binding.jetpackBenefitsBanner.root.isVisible = false
```
I updated this one and the content state too to hide the Jetpack CP banner.

hafizrahman commented:
> when I tested the installation, I noticed that the UI at the step of installation is not accurate,

Thanks, I also just noticed this. Might be a regression somewhere, I'll fix it.

hafizrahman committed:
Fix: Display bugs.

Had to remove the uses of `Group` as introduces odd visibility clashes when multiple groups contain similar referenced ids.

hafizrahman committed:
Fix: Unit test.

hafizrahman commented:
@hichamboushaba nice idea using `Job` and `selectedSite.observe()`.

I think this is good to go and I've added a fix for the UI issue that you mentioned it. Can you give this a quick test, and then we can merge this?

hichamboushaba committed:
Add a missing call to `dropView`

hichamboushaba commented:
This looks good, except that I noticed a crash while testing it:
```
Process: com.woocommerce.android.prealpha, PID: 5678
    java.lang.NullPointerException
        at com.woocommerce.android.ui.prefs.MainSettingsFragment.getBinding(MainSettingsFragment.kt:48)
        at com.woocommerce.android.ui.prefs.MainSettingsFragment.handleJetpackInstallOption(MainSettingsFragment.kt:247)
        at com.woocommerce.android.ui.prefs.MainSettingsPresenter.setupJetpackInstallOption(MainSettingsPresenter.kt:59)
```
The cause is that `MainSettingsFragment` was missing a call to `dropView` which would cancel the presenter's `coroutineScope`, so we kept listening for Jetpack installation changes even when the screen wasn't visible, I fixed this in c9c67f4, since it's my fault.

@JorgeMucientes if you can please review this PR, since me and Hafiz both did changes, and it would be better to have someone else's approval 🙏

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreViewModel.kt:
```diff
@@ -192,6 +190,17 @@ class MyStoreViewModel @Inject constructor(
+            selectedSite.observe()
+                .filter { it?.isJetpackConnected == true }
+                .take(1)
+                .collect { loadStoreStats(activeStatsGranularity.value) }
```
Shouldn't Top Performers be loaded too once Jetpack is connected? 

Non blocking but maybe a unit test that verifies that when selectedSite `isJetpackConnected` changes state we update the UI accordingly.

hichamboushaba replied:
> Shouldn't Top Performers be loaded too once Jetpack is connected?

I don't think so, the top performers is retrieved from the remote site, so having Jetpack or Jetpack the plugin doesn't impact it.
Even the order stats are not impacted, we are reloading them just to refresh the whole stats view, which would allow:
- Hiding the Jetpack CP UI.
- Fetching the visitor stats.

> Non blocking but maybe a unit test that verifies that when selectedSite isJetpackConnected changes state we update the UI accordingly.

Good call, @hafizrahman can you handle this one on this PR, or at least create a ticket to track it for a future PR 🙂

JorgeMucientes replied:
Cool! Just double checking. Good to know Jetpack install only affects visitors stats. 

I've pre-approved the PR, not merging though just in case @hafizrahman wants to add some tests.

hichamboushaba replied:
@hafizrahman just a friendly reminder about this discussion 🙂

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/jetpack/JetpackInstallProgressDialog.kt:
```diff
@@ -112,20 +108,22 @@ class JetpackInstallProgressDialog : DialogFragment(R.layout.dialog_jetpack_inst
         }
     }
 
+    private fun setViewVisibility(visibility: Int, vararg views: View) = views.forEach { it.visibility = visibility }
```
Nice! Did not occur to me to use vararg to set visibility for multiple views at the time.

JorgeMucientes commented:
Nice job folks @hichamboushaba @hafizrahman 👏.  

Code looks good and I think everything works as intended. I just left one comment about refreshing top performers list as well once Jetpack is installed. Let me know what you think.

hafizrahman committed:
Add: Unit test for after JP install.

hafizrahman commented:
Merging now as I've added the suggested unit test :)
